runtime/common/src/main/kotlin/kotlinx/serialization/json/internal/TreeJsonOutput.kt

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright 2018 JetBrains s.r.o.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package kotlinx.serialization.json.internal
+
+import kotlinx.serialization.*
+import kotlinx.serialization.internal.*
+import kotlinx.serialization.json.*
+import kotlin.collections.set
+
+internal fun <T> Json.writeJson(value: T, serializer: SerializationStrategy<T>): JsonElement {
+    lateinit var result: JsonElement
+    val encoder = JsonTreeOutput(this) { result = it }
+    encoder.encode(serializer, value)
+    return result
+}
+
+private sealed class AbstractJsonTreeOutput(
+    override val json: Json,
+    val nodeConsumer: (JsonElement) -> Unit
+) : NamedValueEncoder(), JsonOutput {
+
+    override fun writeTree(tree: JsonElement) {
+        nodeConsumer(tree)
+    }
+
+    override fun shouldEncodeElementDefault(desc: SerialDescriptor, index: Int): Boolean = json.encodeDefaults
+    override fun composeName(parentName: String, childName: String): String = childName
+    abstract fun putElement(key: String, element: JsonElement)
+    abstract fun getCurrent(): JsonElement
+
+    override fun encodeTaggedNull(tag: String) = putElement(tag, JsonNull)
+
+    override fun encodeTaggedInt(tag: String, value: Int) = putElement(tag, JsonLiteral(value))
+    override fun encodeTaggedByte(tag: String, value: Byte) = putElement(tag, JsonLiteral(value))
+    override fun encodeTaggedShort(tag: String, value: Short) = putElement(tag, JsonLiteral(value))
+    override fun encodeTaggedLong(tag: String, value: Long) = putElement(tag, JsonLiteral(value))
+
+    override fun encodeTaggedFloat(tag: String, value: Float) {
+        if (json.strictMode && !value.isFinite()) {
```

Comment:
This check probably could be leveraged to somewhere around `JsonOutput`, since it is used in regular json, too